### PR TITLE
Hash specs cleanup

### PR DESCRIPTION
**Pull request by R1kM**
## Proposed changes

This PR addresses the last remaining comment in #735. It moves the definition of finish to Spec.Agile.Hash, and renames Spec.Hash.PadFinish into Spec.Hash.MD, since it now only contains the definition of pad that is only defined for MD algorithms.
It also removes finish and pad from Spec.SHA1, Spec.SHA2, and Spec.MD5, as their definitions are unused and these functions are only defined through the generic finish and pad definitions.
As a byproduct, it also compacts Spec.Hash.Lemmas0 and Spec.Hash.Lemmas into a single module, moves Spec.Hash.Lemmas to specs/lemmas, and removes a few unused open primitives in different files. 

Note, we should probably also split the implementation file Hacl.Hash.PadFinish following the changes to the specs and rename it accordingly, but this can be done in a further round of cleanups.

This PR does not alter the extracted C code.

## Types of changes

- [x] Refactoring

## Checklist

- [x] I have read the [CODE_OF_CONDUCT.md](CODE_OF_CONDUCT.md) doc
- [x] I have read and agree to submit my changes under the [LICENSE](LICENSE.md)
- [x] I have added necessary documentation (if appropriate)
- [x] I have edited [CHANGES.md](CHANGES.md) (if appropriate)

### CODE REVIEW

File: specs/Spec.Hash.MD.fst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    // Saves the need for high fuel + makes hint replayable.
+    max_input_size_len a;
+    let encodedlen : lbytes (len_length a) =
+      match a with
+      | MD5 -> Lib.ByteSequence.uint_to_bytes_le (secret (nat_to_len a (total_len * 8)))
+      | _ -> Lib.ByteSequence.uint_to_bytes_be (secret (nat_to_len a (total_len * 8)))
+    in
+    S.(firstbyte @| zeros @| encodedlen)
+
+let pad (a:md_alg)
```
Comment (msprotz) [inline specs/Spec.Hash.MD.fst:45]: get rid of this superfluous definition? and rename `pad_md` into `pad`

**Review by msprotz — APPROVED** [review]
Great for me. Thanks!

File: specs/Spec.Hash.MD.fst (hunk 1 above)
Comment (R1kM) [reply specs/Spec.Hash.MD.fst]: Good point, thanks. Done in [79fe881](https://github.com/hacl-star/hacl-star/pull/740/commits/79fe8815183f9132fd38dfa22ec3e3b68cfcd117)

**Review by polubelova — APPROVED** [review]
Thanks!

**Comment by msprotz** [comment]
```
/build/22s5z4675x999l4zz35r23kjh6zyb8wg-source/code/hpke/Hacl.Meta.HPKE.fst(5,0-15,2): (Error 19) could not prove post-condition; Try with --query_stats to get more details (see also /build/22s5z4675x999l4zz35r23kjh6zyb8wg-source/code/hpke/Hacl.Impl.HPKE.fst(760,32-760,77))
```

want me to merge your other cleanup PR first?

**Pull request merged**